### ISSUE TITLE
Add directives to extract the entity and/or the payload (dataBytes)

### ISSUE THREAD

**Issue by hseeberger** [akka/akka#19756]
`BasicDirecteives` defines a lot of `extract` methods, e.g. `extractRequest`. Would it make sense to add another two: `extractEntity` and `extractData` (the latter extracting the payload)? If you agree, I'm happy to provide a PR.

**Comment by rkuhn** [comment]
Looks sensible to me, but might want to be done after #19632 (or take it into account).

**Comment by johanandren** [comment]
Sounds like a good idea to me.

**Comment by hseeberger** [comment]
As discussed on Gitter, one could of course use `extractRequest` and then access the entity or payload via `request.entity` or `request.entity.dataBytes`, but (a) the new directives are more convenient and (b) easier to find, e.g. via code completion.

**Comment by ktoso** [comment]
That's why we have http://doc.akka.io/docs/akka-stream-and-http-experimental/2.0/scala/http/routing-dsl/directives/marshalling-directives/entity.html

**Comment by jrudolph** [comment]
For the sake of completeness it would make sense to add them.

The main point against adding them is that they draw attention away from using unmarshallers and the `entity` directive which is the idiomatic way to access entity data. `entity` will automatically flag errors if it encounters unknown content-types and encapsulates the low-level unmarshalling logic so that you can work with high-level domain objects instead.

**Comment by hseeberger** [comment]
@ktoso, how would `entity` help me when I want to access the data source?

**Comment by hseeberger** [comment]
@jrudolph, same question, more details: If I want to consume the incoming payload, let's say a file upload, via Akka Streams, how would `entity` help me?

**Comment by jrudolph** [comment]
The idiomatic way would be to provide an `Unmarshaller` which returns a value that represents the file upload.

**Comment by hseeberger** [comment]
@jrudolph, so far I have looked at `entity` as a way to bridge from the HTTP model to the domain model. Above you also said "that you can work with high-level domain objects instead".

Yet in this case I'm not interested in any domain objects, I want to work with the payload of type `Source[ByteString, Any]`. Shouldn't that be accessible directly without the `entity`-detour? And with detour I mean unnecessary code/invocations as well as an intellectual detour: From HTTP to domain and back to HTTP.

So I argue that the idiomatic way for my use case is something like `extractRequest { _.entity.dataBytes ... }`. Do you agree with this?

Of course this might be a rare case, hence I'm happy to accept that we should not add these directives.

**Comment by jrudolph** [comment]
I see, you just want to access the data but don't convert it into any domain model or return any result. In this case it would be nice to have such a low-level directive. 

I guess the best practice would be to have those directives but discourage from using them in your main route. Instead you would build your own little set of directives and routes (using those low-level directives) which you would then reference in your main route. 

The insistence on using `entity` is maybe just a lazy way of getting people to actual abstract stuff like that instead of actually educating them about the best practices and the underlying reasons.

**Comment by ktoso** [comment]
Right, I see - I guess it makes sense to make it easy to access the entity (`HttpEntity` not just the dataBytes).
Would be cool if you could PR this addition, thanks @hseeberger !

**Comment by hseeberger** [comment]
@jrudolph, I don't get the point with this "own little set of directives". Well, sometimes and completely independent of these concrete directives that's a good idea. But why should I build some custom directive instead of adding the following to my main route:

``` scala
pathSingleSlash {
  post {
    extractData { data =>
      val path = Files.createTempFile("streamz", "")
      val save = data
        .throttle(100 * 1024, 1.second, 100 * 1024, _.length, ThrottleMode.Shaping)
        .runWith(FileIO.toFile(path.toFile))
      onSuccess(save) {
        case IOResult(count, Success(Done)) => complete(StatusCodes.Created -> s"$path: $count")
        case IOResult(_, Failure(cause))    => complete(StatusCodes.InternalServerError -> cause.getMessage)
      }
    }
  }
}
```

@ktoso, I don't have spare cycles before 2.4.2, but that shouldn't be an issue, right?

**Comment by ktoso** [comment]
Yeah, no problem about missing 2.4.2 (DSLs still experimental for now)

**Comment by ktoso** [comment]
Another directive up for grabs!

**Comment by kwyczesany** [comment]
Hi, I would like to take this task if nobody is working on it

**Comment by ktoso** [comment]
Good pick, let us know if you need any help :)
